File: benchmarks/c/b12_ga.c
/* Genetic Algorithm minimizing the Rosenbrock function —
 * Computational Intelligence benchmark (MICAI). Tournament selection, uniform
 * crossover, uniform mutation. Rosenbrock uses only +,-,* (no cos as in
 * Rastrigin) and mutation is uniform (no Gaussian sqrt/log), so the entire
 * search is bit-exact across IEEE-754 implementations. Every random decision
 * draws from the shared 64-bit LCG (seed 42) in an identical order in all six
 * languages, making the whole evolution deterministic.
 * Checksum = best fitness found + sum of the best individual's genes (6 dp). */
#include <stdio.h>

#define D 30          /* genes per individual */
#define P 5000        /* population size */
#define G 1200        /* generations (sized so C ~1s) */
#define T 3           /* tournament size */
#define MUT_RATE 0.1  /* per-gene mutation probability */
#define MUT_STEP 0.1  /* uniform mutation half-range */

typedef struct { unsigned long long state; } SimpleRNG;

static double next_double(SimpleRNG *rng) {
    rng->state = rng->state * 6364136223846793005ULL + 1442695040888963407ULL;
    return (double)(rng->state >> 33) / (double)(1ULL << 31);
}

static double pop[P][D];
static double newpop[P][D];
static double fitness[P];
static double best_genes[D];

static double rosenbrock(const double *x) {
    double f = 0.0;
    for (int i = 0; i < D - 1; i++) {
        double a = x[i + 1] - x[i] * x[i];
        double b = 1.0 - x[i];
        f += 100.0 * a * a + b * b;
    }
    return f;
}

int main(void) {
    SimpleRNG rng = { .state = 42 };
    for (int p = 0; p < P; p++)
        for (int d = 0; d < D; d++)
            pop[p][d] = (next_double(&rng) * 2.0 - 1.0) * 5.0;  /* genes in [-5, 5] */

    double best_fit = 1e300;
    for (int g = 0; g < G; g++) {
        for (int p = 0; p < P; p++) {
            double f = rosenbrock(pop[p]);
            fitness[p] = f;
            if (f < best_fit) {
                best_fit = f;
                for (int d = 0; d < D; d++) best_genes[d] = pop[p][d];
            }
        }
        for (int d = 0; d < D; d++) newpop[0][d] = best_genes[d];  /* elitism */
        for (int i = 1; i < P; i++) {
            int a = (int)(next_double(&rng) * P);
            for (int t = 1; t < T; t++) {
                int idx = (int)(next_double(&rng) * P);
                if (fitness[idx] < fitness[a]) a = idx;
            }
            int b = (int)(next_double(&rng) * P);
            for (int t = 1; t < T; t++) {
                int idx = (int)(next_double(&rng) * P);
                if (fitness[idx] < fitness[b]) b = idx;
            }
            for (int d = 0; d < D; d++)
                newpop[i][d] = (next_double(&rng) < 0.5) ? pop[a][d] : pop[b][d];
            for (int d = 0; d < D; d++)
                if (next_double(&rng) < MUT_RATE)
                    newpop[i][d] += (next_double(&rng) * 2.0 - 1.0) * MUT_STEP;
        }
        for (int p = 0; p < P; p++)
            for (int d = 0; d < D; d++) pop[p][d] = newpop[p][d];
    }

    double gsum = 0.0;
    for (int d = 0; d < D; d++) gsum += best_genes[d];
    printf("%.6f %.6f\n", best_fit, gsum);
    return 0;
}

File: benchmarks/python/b12_ga.py
# Genetic Algorithm minimizing the Rosenbrock function —
# Computational Intelligence benchmark (MICAI). Tournament selection, uniform
# crossover, uniform mutation. Rosenbrock + uniform mutation use only +,-,*,/,
# so the search is bit-exact across languages. Every random decision draws from
# the shared 64-bit LCG (seed 42) in identical order in all six languages.
# Checksum = best fitness found + sum of the best individual's genes (6 dp).
D = 30
P = 5000
G = 1200
T = 3
MUT_RATE = 0.1
MUT_STEP = 0.1


class SimpleRNG:
    def __init__(self, seed=42):
        self.state = seed

    def next_double(self):
        self.state = (self.state * 6364136223846793005 + 1442695040888963407) & 0xFFFFFFFFFFFFFFFF
        return (self.state >> 33) / (1 << 31)


def rosenbrock(x):
    f = 0.0
    for i in range(D - 1):
        a = x[i + 1] - x[i] * x[i]
        b = 1.0 - x[i]
        f += 100.0 * a * a + b * b
    return f


def main():
    rng = SimpleRNG(42)
    pop = [[(rng.next_double() * 2.0 - 1.0) * 5.0 for _ in range(D)] for _ in range(P)]
    newpop = [[0.0] * D for _ in range(P)]
    fitness = [0.0] * P
    best_genes = [0.0] * D

    best_fit = 1e300
    for g in range(G):
        for p in range(P):
            f = rosenbrock(pop[p])
            fitness[p] = f
            if f < best_fit:
                best_fit = f
                best_genes = pop[p][:]
        newpop[0] = best_genes[:]
        for i in range(1, P):
            a = int(rng.next_double() * P)
            for t in range(1, T):
                idx = int(rng.next_double() * P)
                if fitness[idx] < fitness[a]:
                    a = idx
            b = int(rng.next_double() * P)
            for t in range(1, T):
                idx = int(rng.next_double() * P)
                if fitness[idx] < fitness[b]:
                    b = idx
            ni = newpop[i]
            pa = pop[a]
            pb = pop[b]
            for d in range(D):
                ni[d] = pa[d] if rng.next_double() < 0.5 else pb[d]
            for d in range(D):
                if rng.next_double() < MUT_RATE:
                    ni[d] += (rng.next_double() * 2.0 - 1.0) * MUT_STEP
        # swap populations (copy values to keep pop/newpop distinct buffers)
        for p in range(P):
            pop[p] = newpop[p][:]

    gsum = 0.0
    for d in range(D):
        gsum += best_genes[d]
    print(f"{best_fit:.6f} {gsum:.6f}")


main()
